File: datastructure/week4/stack.c
#include <stdio.h>
#include <stdlib.h>
#define SIZE 100

typedef struct
{
    char data[SIZE];
    int top;
}StackType;

void init(StackType * S)
{   
    S->top = -1; //배열이 아닌 배열의 밖을 지정해주는 행위
}

int isEmpty(StackType * S)
{
    return S->top == -1; //비었는지 안 비었는지 리턴
}

int isFull(StackType * S)
{
    return S->top == SIZE-1; // 꽉 차있는지 
}

void push(StackType * S, char elem)
{
    if(isFull(S))
        printf("Overflow!!\n");
     // 여기서는 주소연산자를 사용 안함 이미 주소를 받았기 때문에
    else
        S->top++;
        S->data[S->top] = elem;
    
}

char pop(StackType * S)
{
    if(isEmpty(S))
       { printf("Empty!!\n");
        return -1;
       }
    char elem = S->data[S->top];
    S->top--;
    return elem;
} //값을 리턴하고 제거

char peek(StackType * S)
{
    if(isEmpty(S))
       { printf("Empty!!\n");
        return -1;
       }
    return S->data[S->top];
} //값만 리턴

void print(StackType * S)
{
    for(int i=0; i<=S->top; i++){
        printf("%c ", S->data[i]);
    }
    printf("\n");
}


int main(void){
    StackType S;
    init(&S);

    pop(&S); getchar(); // 오류를 체크하기 위해
    push(&S, 'c');
    push(&S, 'a');
    push(&S, 't');
    push(&S, 's');
    print(&S); getchar();

    printf("Call pop() : %c\n", pop(&S)); print(&S); getchar();
    printf("Call peek() : %c\n", peek(&S)); print(&S); getchar();
    return 0;
}

File: datastructure/Java/stack.java
import java.util.*;

public class stack{
    public static final int SIZE = 1000;
    public 
        char data[] = new char [SIZE];
        int point = 0;
    public stack(){
        this.point = -1;
    }

    public void init(){
        this.point = -1;
    }

    public boolean isFull(){
        if(this.point == (SIZE-1)){
            return true;
        }
        else return false;
    }

    public boolean isEmpty(){
        if(this.point == -1){
            return true;
        }
        else{
            return false;
        }
    }
    public void push(char elem){
        if(this.isFull()){
            System.out.println("This stack is Full !!!!\n");
        }
        else{
            this.point++;
            this.data[this.point] = elem;
        }
    }

    public char pop(){
        if(this.isEmpty()){
            System.out.println("This stack is Empty!!!!\n");
            return 'e';
        }
        else{
            char temp = this.data[this.point];
            this.point = this.point -1;
            return temp;
        }
    }

    public char peek(){
        if(this.isEmpty()){
            System.out.println("This stack is Empty!!!!\n");
            return 'e';
        }
        else{
            return this.data[this.point];
        }
    }

    public void print(){
        for(int i=0; i<=this.point; i++){
            System.out.print(this.data[i]+" ");
        }
        System.out.println();
    }
    public static void main(String[] args){
        stack s = new stack();
        s.init();
        s.push('A');
        s.print();
        System.out.println(s.peek());
        System.out.println(s.pop());
    }
}

